src/frontend.rs

```diff
@@ -136,207 +140,333 @@ pub struct AnchoredKey(pub Origin, pub String);
+        // TODO: argument type checking
+        let st = match op_type {
+            None => Statement(NativeStatement::None, vec![]),
+            NewEntry => Statement(NativeStatement::ValueOf, self.op_args_entries(args)),
+            CopyStatement => todo!(),
+            EqualFromEntries => Statement(NativeStatement::Equal, self.op_args_entries(args)),
+            NotEqualFromEntries => Statement(NativeStatement::NotEqual, self.op_args_entries(args)),
+            GtFromEntries => Statement(NativeStatement::Gt, self.op_args_entries(args)),
+            LtFromEntries => Statement(NativeStatement::Lt, self.op_args_entries(args)),
+            TransitiveEqualFromStatements => todo!(),
+            GtToNotEqual => todo!(),
+            LtToNotEqual => todo!(),
+            ContainsFromEntries => Statement(NativeStatement::Contains, self.op_args_entries(args)),
+            NotContainsFromEntries => {
+                Statement(NativeStatement::NotContains, self.op_args_entries(args))
+            }
+            RenameContainedBy => todo!(),
+            SumOf => todo!(),
+            ProductOf => todo!(),
+            MaxOf => todo!(),
+        };
+        self.operations.push(op);
+        self.statements.push(st);
+        &self.statements[self.statements.len() - 1]
+    }
+
+    pub fn prove<P: PodProver>(&self, prover: &mut P) -> MainPod {
+        let compiler = MainPodCompiler::new(&self.params);
+        let inputs = MainPodCompilerInputs {
+            // signed_pods: &self.input_signed_pods,
+            // main_pods: &self.input_main_pods,
+            statements: &self.statements,
+            operations: &self.operations,
+        };
+        let (statements, operations) = compiler.compile(inputs).expect("TODO");
+
+        let inputs = middleware::MainPodInputs {
+            signed_pods: &self
+                .input_signed_pods
+                .iter()
+                .map(|p| p.pod.as_ref())
+                .collect_vec(),
+            main_pods: &self
+                .input_main_pods
+                .iter()
+                .map(|p| p.pod.as_ref())
+                .collect_vec(),
+            statements: &statements,
+            operations: &operations,
+        };
+        let pod = prover.prove(&self.params, inputs);
+        MainPod { pod }
     }
 }
 
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct MainPod {
-    pub params: Params,
-    pub id: PodId,
-    pub input_signed_pods: Vec<SignedPod>,
-    pub input_main_pods: Vec<MainPod>,
-    pub statements: Vec<(Statement, Operation)>,
+    pub pod: Box<dyn middleware::MainPod>,
+    // TODO: metadata
 }
 
 impl MainPod {
-    pub fn origin(&self) -> Origin {
-        Origin(PodType::Main, self.id)
+    pub fn id(&self) -> PodId {
+        self.pod.id()
     }
-
-    pub fn compile(&self) -> Result<backend::MainPod> {
-        MainPodCompiler::new(self).compile()
+    pub fn origin(&self) -> Origin {
+        Origin(PodClass::Signed, self.id())
```

Comment:
should this be PodClass::Main?

Reply:
yes!

Reply:
I've pushed the fix in this PR https://github.com/0xPARC/pod2/pull/23
https://github.com/0xPARC/pod2/pull/23/commits/7dc9597de5d42195ccfdcc4a97e4e61b8297f4c4